056-zk-kraft-migration.md

```diff
@@ -0,0 +1,258 @@
+
+The operator has to wait for the KRaft controllers being ready, the quorum to be formed and leader elected, then the migration process moves to the state of waiting for brokers to register.
+
+The FSM moves to a `WaitBrokersInMigration` state where the operator expects the user to ask for running the actual metadata migration.
+It means the operator has to enable the ZooKeeper migration on the brokers pool in order to actually start the migration (Phase 2).
+
+From this state it's still possible to not move forward with the migration.
+The user can delete the `KafkaNodePool` custom resource containing the definition of the KRaft controllers and remove the `strimzi.io/kraft-migration` annotation from the `Kafka` custom resource.
+The FSM moves back to the `NoMigration` state (or null) actually exiting from the migration process and the operator removes the `strimzi.io/kraft-migration-state` annotation from the `Kafka` custom resource.
+There is nothing to do with the brokers which are still configured to work with the ZooKeeper ensemble only.
+
+#### Phase 2: Enabling the migration on the brokers
+
+In the `WaitBrokersInMigration` state, the operator expects the user to ask for running the actual metadata migration.
+In order to start the actual migration, the brokers need to be connected to the KRaft controllers and have the ZooKeeper migration enabled.
+The user annotates the `Kafka` custom resource running the current ZooKeeper based brokers with `strimzi.io/kraft-migration: migration`.
+This annotation drives the operator to update the brokers' configuration with the ZooKeeper migration flag enabled and with the new KRaft controllers quorum connection information as well.
+
+```shell
+# KRaft controller quorum configuration
+controller.quorum.voters=<voters>
+controller.listener.names=CONTROLLER
+
+# The controller.listener.name should also be added to listener.security.property.map
+listener.security.protocol.map=PLAINTEXT:PLAINTEXT,CONTROLLER:PLAINTEXT
+
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+```
+
+Finally, the operator rolls the brokers.
+After the rolling, the internal Apache Kafka migration component starts the actual metadata migration and the operator has to wait for it to be completed to move to the next phase.
+Checking the migration progress can be done using some of the exposed migration metrics.
+The FSM moves to `Migration` state first, while migration is ongoing, and finally to the `DualWrite` state.
+In both states, the user can rollback the process by applying the `strimzi.io/kraft-migration: pre-migration` on the `Kafka` custom resource.
+In this case, the operator configures the brokers with no ZooKeeper migration flag, and no KRaft controllers quorum connection information.
+Finally, the operator rolls the brokers and wait for them to be ready again.
+The FSM moves back to the `WaitBrokersInMigration`, so actually at the beginning of the Phase 2.
```

Comment:
Does it mean that from this stage, you cannot go directly to the beginning and just remove the controller nodes?

Reply:
Well it should be semantically and technically possible. If the user deletes the `KafkaNodePool` for controllers in this state, the operator detects it, can roll brokers to come back using ZooKeeper and revert the migration.
Is this what you meant? Instead of applying the `pre-migration` annotation? Maybe we can leave both approach?

Reply:
Well, it sounded like you suggest you can move only one step back but say nothing abotu moving two steps back. So I was curious.

Reply:
Well that was a good point. I was envisaging user to revert back brokers first but not removing controllers yet. Maybe changing his mind and again restarting migration :-)
Anyway we could have both if you like.